// @entria/generator-graphql/generators/add/index.js
'use strict';

var _yeomanGenerator = require('yeoman-generator');

var _yeomanGenerator2 = _interopRequireDefault(_yeomanGenerator);

var _pluralize = require('pluralize');

var _pluralize2 = _interopRequireDefault(_pluralize);

var _utils = require('../utils');

var _config = require('../config');

var _ejsHelpers = require('../ejsHelpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class AddGenerator extends _yeomanGenerator2.default {
  constructor(args, options) {
    super(args, options);

    this.argument('name', {
      type: String,
      required: true
    });

    // TODO read schema.json

    this.destinationDir = (0, _config.getConfigDir)('add');
  }

  _getConfigDirectories() {
    return (0, _utils.getRelativeConfigDir)('loader', ['model', 'connection']);
  }

  generateList() {
    // const schema = this.options.model ?
    //   getMongooseModelSchema(this.options.model, true)
    //   : null;

    const name = (0, _ejsHelpers.uppercaseFirstLetter)(this.options.name);

    // const templatePath = schema ?
    //   this.templatePath('LoaderWithSchema.js.template')
    //   : this.templatePath('Loader.js.template');
    //
    // const directories = this._getConfigDirectories();

    const pluralName = (0, _pluralize2.default)(this.options.name);

    const templateVars = {
      name,
      rawName: this.options.name,
      camelCaseName: (0, _ejsHelpers.camelCaseText)(name),
      pluralName,
      pluralCamelCaseName: (0, _ejsHelpers.camelCaseText)(pluralName)
    };

    const files = {
      add: {
        filename: `${name}Add.js`,
        template: 'Add.js.template'
      },
      addMutation: {
        filename: `${name}AddMutation.js`,
        template: 'AddMutation.js.template'
      }
    };

    Object.keys(files).forEach(file => {
      const { filename, template } = files[file];

      this.fs.copyTpl(this.templatePath(template), `${this.destinationDir}/${filename}`, templateVars);
    });
  }

  end() {
    this.log('🔥 Add created!');
  }
}

module.exports = AddGenerator;
