// @entria/generator-graphql/generators/list/index.js
'use strict';

var _yeomanGenerator = require('yeoman-generator');

var _yeomanGenerator2 = _interopRequireDefault(_yeomanGenerator);

var _pluralize = require('pluralize');

var _pluralize2 = _interopRequireDefault(_pluralize);

var _config = require('../config');

var _ejsHelpers = require('../ejsHelpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ListGenerator extends _yeomanGenerator2.default {
  constructor(args, options) {
    super(args, options);

    this.argument('name', {
      type: String,
      required: true
    });

    // TODO read schema.json

    this.destinationDir = (0, _config.getConfigDir)('list');
  }

  generateList() {
    const name = (0, _ejsHelpers.uppercaseFirstLetter)(this.options.name);

    const templatePath = this.templatePath('List.js.template');

    const pluralName = (0, _pluralize2.default)(this.options.name);

    const destinationPath = this.destinationPath(`${this.destinationDir}/${name}List.js`);

    const templateVars = {
      name,
      pluralName
    };

    this.fs.copyTpl(templatePath, destinationPath, templateVars);
  }

  end() {
    this.log('🔥 List created!');
  }
}

module.exports = ListGenerator;
