// @entria/generator-graphql/generators/connection/index.js
'use strict';

var _yeomanGenerator = require('yeoman-generator');

var _yeomanGenerator2 = _interopRequireDefault(_yeomanGenerator);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _config = require('../config');

var _ejsHelpers = require('../ejsHelpers');

var _paths = require('../paths');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class ConnectionGenerator extends _yeomanGenerator2.default {
  constructor(args, options) {
    super(args, options);

    this.argument('name', {
      type: String,
      required: true
    });

    this.destinationDir = (0, _config.getConfigDir)('connection');
  }

  generateConnection() {
    const name = (0, _ejsHelpers.uppercaseFirstLetter)(this.options.name);

    const templatePath = this.templatePath('Connection.js.template');

    const moduleName = this.options.name.toLowerCase();
    const modulePath = (0, _paths.getModulePath)(this.destinationDir, moduleName);

    const destinationPath = this.destinationPath(_path2.default.join(modulePath, `${name}Connection.js`));

    const templateVars = {
      name
    };

    this.fs.copyTpl(templatePath, destinationPath, templateVars);
  }

  end() {
    this.log('🔥 Connection created!');
  }
}


module.exports = ConnectionGenerator;
